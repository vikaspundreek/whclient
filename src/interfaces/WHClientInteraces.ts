export interface IWHConfig {
  apiEndPoint: string;
  tabs: IWHClientTab[];
}
export interface IWHClientTab {
  id: string;
  label: string;
}

export interface IProduct {
  id: string;
  name: string;
  articles: IArticle[];
  stock?: number;
}

export interface IArticle {
  id: string;
  name?: string;
  amountInStock?: number;
  amountRequired: number;
}

export interface ISale {
  id: string;
  createdAt: string;
  productId: string;
  productName?: string;
  amountSold: number;
}

export interface IStore {
  products: IProduct[];
  articles: IArticle[];
  sales: ISale[] | null;
}

export interface IBulkArticleUpdate {
  id: string;
  amountToSubtract: number;
}
