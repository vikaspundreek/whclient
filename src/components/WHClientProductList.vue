<template>
  <div class="wh-client__product-list">
    <div
      class="wh-client__product pt-2"
      v-for="product in products"
      :key="product.index"
    >
      <h6>{{ product.name }}</h6>
      <WHClientArticleList
        v-if="articlesDataAvailable(product)"
        :articles="product.articles"
      />
      <p v-else>
        <strong> Articles </strong>: articles detail not available at this
        moment.
      </p>
      <hr />
    </div>
  </div>
</template>
<script lang="ts">
import { Component, Prop, Vue } from "vue-property-decorator";
import { IArticle, IProduct } from "../interfaces/WHClientInteraces";
import WHClientArticleList from "./WHClientArticleList.vue";

@Component({
  components: { WHClientArticleList },
})
export default class WHClientProductList extends Vue {
  @Prop() private products!: IProduct[];

  private articlesDataAvailable(product: IProduct) {
    return (
      product.articles.length > 0 &&
      product.articles.every((article: IArticle) => article.name !== undefined)
    );
  }
}
</script>
