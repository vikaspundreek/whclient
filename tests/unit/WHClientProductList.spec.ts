import { mount } from "@vue/test-utils";
import WHClientProductList from "@/components/WHClientProductList.vue";
import { IProduct } from "@/interfaces/WHClientInteraces";

const mockProductData: IProduct[] = [
  {
    id: "a269a247-0d38-4b47-9630-79c9ae545b68",
    name: "Dining Chair",
    articles: [
      {
        id: "0517f083-0e15-4876-8d1f-6fa45900431c",
        amountRequired: 4,
        name: "Leg",
        amountInStock: 12,
      },
      {
        id: "831b92b8-677b-42cc-a585-335ea4ccccb6",
        amountRequired: 1,
        name: "Seat",
        amountInStock: 2,
      },
      {
        id: "addc65a8-c759-41d8-a18a-89fe446ad484",
        amountRequired: 8,
        name: "Screw",
        amountInStock: 17,
      },
    ],
  },
  {
    id: "6fed6191-ee01-4563-a33d-5010abe0db36",
    name: "Dining Table",
    articles: [
      {
        id: "0517f083-0e15-4876-8d1f-6fa45900431c",
        amountRequired: 4,
        name: "Leg",
        amountInStock: 12,
      },
      {
        id: "6892b98b-9b87-4520-9a9e-7528f1d78cb4",
        amountRequired: 1,
        name: "Table Top",
        amountInStock: 0,
      },
      {
        id: "addc65a8-c759-41d8-a18a-89fe446ad484",
        amountRequired: 8,
        name: "Screw",
        amountInStock: 17,
      },
    ],
  },
];

describe("WHClientProductList.vue", () => {
  it("Render product list as per provided products data with articles with name", () => {
    const wrapper = mount(WHClientProductList, {
      propsData: { products: mockProductData },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it("Render product list with one of them having not full article data", () => {
    mockProductData[0].articles[0].name = undefined;
    const wrapper = mount(WHClientProductList, {
      propsData: { products: mockProductData },
    });
    expect(wrapper).toMatchSnapshot();
  });
});
