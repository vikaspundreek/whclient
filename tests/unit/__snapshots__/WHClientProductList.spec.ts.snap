// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WHClientProductList.vue Render product list as per provided products data with articles with name 1`] = `
<div class="wh-client__product-list">
  <div class="wh-client__product pt-2">
    <h6>Dining Chair</h6>
    <table class="table">
      <thead>
        <tr>
          <th scope="col">Article Name</th>
          <th scope="col" class="text-right">Required</th>
          <th scope="col" class="text-right">Availability</th>
        </tr>
      </thead>
      <tbody>
        <tr class="wh-client__article pt-2">
          <td>Leg</td>
          <td class="text-right">4</td>
          <td class="text-right wh-client__stock"><span class="article-stock">12</span></td>
        </tr>
        <tr class="wh-client__article pt-2">
          <td>Seat</td>
          <td class="text-right">1</td>
          <td class="text-right wh-client__stock"><span class="article-stock">2</span></td>
        </tr>
        <tr class="wh-client__article pt-2">
          <td>Screw</td>
          <td class="text-right">8</td>
          <td class="text-right wh-client__stock"><span class="article-stock">17</span></td>
        </tr>
      </tbody>
    </table>
    <hr>
  </div>
  <div class="wh-client__product pt-2">
    <h6>Dining Table</h6>
    <table class="table">
      <thead>
        <tr>
          <th scope="col">Article Name</th>
          <th scope="col" class="text-right">Required</th>
          <th scope="col" class="text-right">Availability</th>
        </tr>
      </thead>
      <tbody>
        <tr class="wh-client__article pt-2">
          <td>Leg</td>
          <td class="text-right">4</td>
          <td class="text-right wh-client__stock"><span class="article-stock">12</span></td>
        </tr>
        <tr class="wh-client__article pt-2">
          <td>Table Top</td>
          <td class="text-right">1</td>
          <td class="text-right wh-client__stock"><span class="article-stock article-stock--unavailable">N/A</span></td>
        </tr>
        <tr class="wh-client__article pt-2">
          <td>Screw</td>
          <td class="text-right">8</td>
          <td class="text-right wh-client__stock"><span class="article-stock">17</span></td>
        </tr>
      </tbody>
    </table>
    <hr>
  </div>
</div>
`;

exports[`WHClientProductList.vue Render product list with one of them having not full article data 1`] = `
<div class="wh-client__product-list">
  <div class="wh-client__product pt-2">
    <h6>Dining Chair</h6>
    <p><strong> Articles </strong>: articles detail not available at this
      moment.
    </p>
    <hr>
  </div>
  <div class="wh-client__product pt-2">
    <h6>Dining Table</h6>
    <table class="table">
      <thead>
        <tr>
          <th scope="col">Article Name</th>
          <th scope="col" class="text-right">Required</th>
          <th scope="col" class="text-right">Availability</th>
        </tr>
      </thead>
      <tbody>
        <tr class="wh-client__article pt-2">
          <td>Leg</td>
          <td class="text-right">4</td>
          <td class="text-right wh-client__stock"><span class="article-stock">12</span></td>
        </tr>
        <tr class="wh-client__article pt-2">
          <td>Table Top</td>
          <td class="text-right">1</td>
          <td class="text-right wh-client__stock"><span class="article-stock article-stock--unavailable">N/A</span></td>
        </tr>
        <tr class="wh-client__article pt-2">
          <td>Screw</td>
          <td class="text-right">8</td>
          <td class="text-right wh-client__stock"><span class="article-stock">17</span></td>
        </tr>
      </tbody>
    </table>
    <hr>
  </div>
</div>
`;
